### PR TITLE
Refactoring the load balancer with an overflow queue to reduce activation waiting time

### PR DESCRIPTION
**Pull request by Ahmed-elsayed-mahmoud**
## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#3755)

## My changes affect the following components

- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [x] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [x] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by tysonnorris** [comment]
Can you provide a description with highlights of interesting code? 
How does this relate (if at all) to #2932 ?

File: tests/src/test/scala/system/basic/WskRestBasicSwift41Tests.scala~7e81011234d9c68ef9de890ca8191cd69823e47e [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (tysonnorris) [inline tests/src/test/scala/system/basic/WskRestBasicSwift41Tests.scala~7e81011234d9c68ef9de890ca8191cd69823e47e]: were these files intentionally added?

File: gradle/docker.gradle [hunk 2]
```diff
@@ -42,7 +42,7 @@ ext {
     dockerRegistry = project.hasProperty('dockerRegistry') ? dockerRegistry + '/' : ''
     dockerImageTag = project.hasProperty('dockerImageTag') ? dockerImageTag : 'latest'
     dockerImagePrefix = project.hasProperty('dockerImagePrefix') ? dockerImagePrefix : 'whisk'
-    dockerTimeout = project.hasProperty('dockerTimeout') ? dockerTimeout.toInteger() : 840
+    dockerTimeout = project.hasProperty('dockerTimeout') ? dockerTimeout.toInteger() : 8400
```
Comment (tysonnorris) [inline gradle/docker.gradle]: why is this needed?

File: core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala [hunk 3]
```diff
@@ -74,6 +73,9 @@ class ShardingContainerPoolBalancer(config: WhiskConfig, controllerInstance: Ins
   private val activationsPerNamespace = TrieMap[UUID, LongAdder]()
   private val totalActivations = new LongAdder()
 
+  var localOverflowActivationCount: Int = 0
```
Comment (tysonnorris) [inline core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: If the topic receives messages from multiple controllers, this count won't reflect whether there are messages left in the topic, I think? For example, if another controller consumes ALL the overflow messages, will this controller be left indefinitely with localOverflowActivationCount > 0?

File: core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala [hunk 4]
```diff
@@ -133,38 +135,91 @@ class ShardingContainerPoolBalancer(config: WhiskConfig, controllerInstance: Ins
...
+      // In case of the system isn't in overflow state or it's in the overflow state but the message is rescheduled
+      // we want to call the schedule algorithm to choose an invoker
+      if (!overflowState.get() || (overflowState.get() && isRescheduled)) {
+        val homeInvoker = hash % invokersToUse.size
+        val stepSize = stepSizes(hash % stepSizes.size)
+        ShardingContainerPoolBalancer.schedule(invokersToUse, schedulingState.invokerSlots, homeInvoker, stepSize)
+      } else {
+        // In case of the system is in overflow state and the ActivationMessage isn't rescheduled
+        // return -1 indicates as an index for the invoker
+        Some(new InstanceId(-1))
```
Comment (tysonnorris) [inline core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: Is this any different than returning `None`?

Comment (markusthoemmes) [reply core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: Echoing @tysonnorris' argument: I think we should refactor the `schedule` method to return `Some` for a found invoker and `None` for no invoker found. The special case of `no invokers available` could arguably either end up in queueing straight away or rejecting the request. If we want to continue to do the latter, let's move the emptiness check out of `schedule`.

File: tests/src/test/scala/system/basic/WskRestBasicSwift41Tests.scala~7e81011234d9c68ef9de890ca8191cd69823e47e (hunk 1 above)
Comment (Ahmed-elsayed-mahmoud) [reply tests/src/test/scala/system/basic/WskRestBasicSwift41Tests.scala~7e81011234d9c68ef9de890ca8191cd69823e47e]: I didn't add this file. I think it is a garbage or temp file.

File: gradle/docker.gradle (hunk 2 above)
Comment (Ahmed-elsayed-mahmoud) [reply gradle/docker.gradle]: I added this because my network is not good and during the installation, it times out very quick. Therefore, I increased the timeout interval.

File: core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala (hunk 4 above)
Comment (Ahmed-elsayed-mahmoud) [reply core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: Yes. We need to differentiate between three cases. First, there is no invokers in the system and this case returns `None`. Second, there is invokers in the system but they are busy not free to schedule and this case returns `Some(new InstanceId(-1))` where -1 is a dummy index. Third, there is invokers in the system and there exists one invoker that can schedule the activation and this case returns `Some(invoker.id)`.

File: gradle/docker.gradle (hunk 2 above)
Comment (markusthoemmes) [reply gradle/docker.gradle]: You can provide this value for you locally, if your network is really super bad. No need to change the default imo.

`gradle -PdockerTimeout=8400 distDocker` for example.

**Comment by markusthoemmes** [comment]
@Ahmed-elsayed-mahmoud any news on this one?

**Comment by markusthoemmes** [comment]
Assigning back to @Ahmed-elsayed-mahmoud. Please give an update on the status of your work on this.

**Comment by markusthoemmes** [comment]
Closing this for now, please reopen if you intend to finish the work on this.
